pkgs/by-name/ku/kubectl-rook-ceph/package.nix

```
@@ -0,0 +1,32 @@
+{
+  buildGoModule,
+  fetchFromGitHub,
+  lib,
+}:
+
+buildGoModule (finalAttrs: {
+  pname = "kubectl-rook-ceph";
+  version = "0.9.3";
+
+  src = fetchFromGitHub {
+    owner = "rook";
+    repo = "kubectl-rook-ceph";
+    tag = "v${finalAttrs.version}";
+    hash = "sha256-stWuRej3ogGETLzVabMRfakoK358lJbK56/hjBh2k2M=";
+  };
+
+  vendorHash = "sha256-fB3S946nv1uH9blek6w2EmmYYcdnBcEbmYELfPH9A04=";
+
+  postInstall = ''
+    ln -s $out/bin/cmd $out/bin/kubectl-rook-ceph
+  '';
```

it provides shell completions, so they should be added:
```suggestion
  nativeBuildInputs = [ installShellFiles ];

  postInstall = 
    ''
      ln -s $out/bin/cmd $out/bin/kubectl-rook-ceph
    '' 
    + lib.optionalString (stdenv.hostPlatform.emulatorAvailable buildPackages) (
      let
        emulator = stdenv.hostPlatform.emulator buildPackages;
      in
      ''
        installShellCompletion --cmd kubectl-rook-ceph \
        --bash <(${emulator} $out/bin/kubectl-rook-ceph --bash-completion) \
        --fish <(${emulator} $out/bin/kubectl-rook-ceph --fish-completion) \
        --zsh <(${emulator} $out/bin/kubectl-rook-ceph --zsh-completion)
      ''
    );
```

---

Okay, that's an interesting way to generate completions. Does it do more than https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/misc/starship/default.nix#L53-L58 ?

---

It allows for generating the completions on platforms that can run the package thru an emulator but not necessary natively run the package. It's also whats recommend by the [nixpkgs manual](https://nixos.org/manual/nixpkgs/unstable/#installshellfiles-installshellcompletion-exampleusage-guarded)

---

Thank you for showing this, i had no idea :) I tried the following but found that kubectl-rook-ceph requires a valid kubeconfig for generating shell completions..

 ```
 nativeBuildInputs = [ installShellFiles ];
  postInstall =
    ''
      ln -s $out/bin/cmd $out/bin/kubectl-rook-ceph
    ''
    + lib.optionalString (stdenv.hostPlatform.emulatorAvailable buildPackages) (
      let
        emulator = stdenv.hostPlatform.emulator buildPackages;
      in
      ''
        installShellCompletion --cmd kubectl-rook-ceph \
        --bash <(${emulator} $out/bin/kubectl-rook-ceph completion bash) \
        --fish <(${emulator} $out/bin/kubectl-rook-ceph completion fish) \
        --zsh <(${emulator} $out/bin/kubectl-rook-ceph completion zsh)
      ''
    );
```

Build failure complaining about "invalid configuration":
```
> Running phase: installPhase
> Error: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
> Error: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
> Error: invalid configuration: no configuration has been provided, try setting KUBERNETES_MASTER environment variable
> ERROR: installShellCompletion: installShellCompletion: installed shell completion file `/nix/store/nrs3qr4sl2gwijg06i0b6ld67yxma98y-kubectl-rook-ceph-0.9.3/share/bash-completion/completions/kubectl-rook-ceph.bash' does not exist or has zero size
```
https://github.com/rook/kubectl-rook-ceph/issues/353

I'm not sure if it's possible to generate the completion commands in this case?

---

hmm I guess we can add them once https://github.com/rook/kubectl-rook-ceph/issues/353 is resolved. In the meantime it's fine not to have them.

---

I've added the build instructions in the code but commented out, let me know if this is a good idea or if it should just be removed.